File: agents/severance/pilot/src/agent/a2a-compat.ts
/**
 * Nasiko speaks A2A 1.0 method names; accept both 1.0 and 0.3 forms.
 */
export const METHOD_MAP: Record<string, string> = {
  SendMessage: 'message/send',
  SendStreamingMessage: 'message/stream',
  GetTask: 'tasks/get',
  CancelTask: 'tasks/cancel',
  TaskResubscription: 'tasks/resubscribe',
};

export const ROLE_MAP: Record<string, string> = {
  ROLE_USER: 'user',
  ROLE_AGENT: 'agent',
};

function normalizeMessage(msg: Record<string, unknown>): void {
  const role = msg.role;
  if (typeof role === 'string' && role in ROLE_MAP) msg.role = ROLE_MAP[role];
  const parts = msg.parts;
  if (!Array.isArray(parts)) return;
  for (const part of parts) {
    if (!part || typeof part !== 'object' || Array.isArray(part)) continue;
    const p = part as Record<string, unknown>;
    if ('kind' in p) continue;
    if ('text' in p) p.kind = 'text';
    else if ('data' in p) p.kind = 'data';
    else if ('file' in p) p.kind = 'file';
  }
}

export function rewriteRpc(payload: unknown): unknown {
  if (!payload || typeof payload !== 'object' || Array.isArray(payload)) return payload;
  const body = payload as Record<string, unknown>;
  const method = body.method;
  if (typeof method === 'string' && method in METHOD_MAP) body.method = METHOD_MAP[method];
  const params = body.params;
  if (params && typeof params === 'object' && !Array.isArray(params)) {
    const message = (params as Record<string, unknown>).message;
    if (message && typeof message === 'object' && !Array.isArray(message)) {
      normalizeMessage(message as Record<string, unknown>);
    }
  }
  return body;
}

export function extractText(message: Record<string, unknown> | undefined): string {
  if (!message) return '';
  const parts = message.parts;
  if (!Array.isArray(parts)) return '';
  const chunks: string[] = [];
  for (const part of parts) {
    if (!part || typeof part !== 'object') continue;
    const p = part as Record<string, unknown>;
    if (typeof p.text === 'string') chunks.push(p.text);
    else if (p.data && typeof p.data === 'object') chunks.push(JSON.stringify(p.data));
  }
  return chunks.join('\n').trim();
}

export function extractNasikoToken(
  headers: Record<string, string | string[] | undefined>,
): string | undefined {
  const raw = headers['x-nasiko-agent-token'] ?? headers['X-Nasiko-Agent-Token'];
  if (Array.isArray(raw)) return raw[0];
  return typeof raw === 'string' ? raw : undefined;
}

export type TaskState =
  | 'submitted'
  | 'working'
  | 'input-required'
  | 'completed'
  | 'failed'
  | 'canceled';

export function makeTask(opts: {
  id: string;
  contextId: string;
  state: TaskState;
  message?: string;
  artifacts?: { name: string; text?: string; data?: unknown }[];
}): Record<string, unknown> {
  const artifacts = (opts.artifacts ?? []).map((a) => {
    const parts: Record<string, unknown>[] = [];
    if (a.data !== undefined) {
      parts.push({ kind: 'data', data: a.data });
      parts.push({
        kind: 'text',
        text: typeof a.text === 'string' ? a.text : JSON.stringify(a.data),
      });
    } else if (a.text !== undefined) {
      parts.push({ kind: 'text', text: a.text });
    }
    return { name: a.name, parts };
  });
  return {
    id: opts.id,
    contextId: opts.contextId,
    status: {
      state: opts.state,
      message: opts.message
        ? { role: 'agent', parts: [{ kind: 'text', text: opts.message }] }
        : undefined,
    },
    artifacts,
    kind: 'task',
  };
}

File: agents/severance/surveyor/src/a2a_compat.py
"""Nasiko's control plane speaks A2A 1.0 method names; a2a-sdk 0.3.26 speaks 0.3."""

from __future__ import annotations

import asyncio
import json

METHOD_MAP = {
    "SendMessage": "message/send",
    "SendStreamingMessage": "message/stream",
    "GetTask": "tasks/get",
    "CancelTask": "tasks/cancel",
    "TaskResubscription": "tasks/resubscribe",
}

ROLE_MAP = {
    "ROLE_USER": "user",
    "ROLE_AGENT": "agent",
}


def _normalize_message(msg: dict) -> None:
    role = msg.get("role")
    if isinstance(role, str) and role in ROLE_MAP:
        msg["role"] = ROLE_MAP[role]
    parts = msg.get("parts")
    if not isinstance(parts, list):
        return
    for part in parts:
        if not isinstance(part, dict):
            continue
        if "kind" in part:
            continue
        if "text" in part:
            part["kind"] = "text"
        elif "data" in part:
            part["kind"] = "data"
        elif "file" in part:
            part["kind"] = "file"


def rewrite_rpc_bytes(body: bytes) -> bytes:
    try:
        payload = json.loads(body.decode() or "null")
    except (UnicodeDecodeError, json.JSONDecodeError):
        return body
    if not isinstance(payload, dict):
        return body
    method = payload.get("method")
    mapped = METHOD_MAP.get(method) if isinstance(method, str) else None
    if mapped:
        payload["method"] = mapped
    params = payload.get("params")
    if isinstance(params, dict) and isinstance(params.get("message"), dict):
        _normalize_message(params["message"])
    return json.dumps(payload).encode()


class A2a10MethodCompat:
    def __init__(self, app):
        self.app = app

    async def __call__(self, scope, receive, send):
        if scope["type"] != "http" or scope.get("method") != "POST":
            await self.app(scope, receive, send)
            return
        chunks = bytearray()
        more = True
        while more:
            event = await receive()
            etype = event.get("type")
            if etype == "http.request":
                chunks.extend(event.get("body") or b"")
                more = bool(event.get("more_body"))
            elif etype == "http.disconnect":
                return
            else:
                more = False
        rewritten = rewrite_rpc_bytes(bytes(chunks))
        headers = []
        for key, value in scope.get("headers") or []:
            if key == b"content-length":
                headers.append((key, str(len(rewritten)).encode()))
            else:
                headers.append((key, value))
        new_scope = dict(scope)
        new_scope["headers"] = headers
        sent = False
        hold = asyncio.Event()

        async def new_receive():
            nonlocal sent
            if not sent:
                sent = True
                return {"type": "http.request", "body": rewritten, "more_body": False}
            # Do not fake a client disconnect — that aborts SSE streams.
            await hold.wait()
            return {"type": "http.disconnect"}

        await self.app(new_scope, new_receive, send)
